File: tempIdeas/java/src/Osso.java
public class Osso extends Oggetto {
    /* parameters inside stats in Pietra are:
     * spirito
     * attacco
     * difesa
     */
    private String attacco_;
    private String difesa_;


    //COSTRUTTORI
    public Osso() {
        this(1,1,1.0,1.0,1.0);
    }

    public Osso(Integer livello, Integer rarita, Double spirito, Double attacco, Double difesa) {
        super(livello, rarita, spirito);
        attacco_="Attacco";
        difesa_="Difesa";
        insertStat(attacco_, attacco);
        insertStat(difesa_, difesa);
        sanitizeInput();
    }

    public Double getAttacco() {
        return getValoreStat(attacco_);
    }
    public Double getDifesa() {
        return getValoreStat(difesa_);
    }

    @Override
    public Osso clone() {
        return new Osso(getLivello(), getRarita(), getSpirito(), getAttacco(), getDifesa());
    }

    public Double ricicla() {
        return calcolaMana()/2 + (getAttacco()+getDifesa())/ 2 * getRarita();
    }

    @Override
    public void potenzia(Double mana) {
        Double incremento=mana/(getLivello()*getListaStats().size());
        getListaStats().stream().forEach(s->incrementStat(s, incremento));

        Integer divisore =5;
        incrementStat(attacco_, incremento*getRarita()/(2*divisore));
        incrementStat(difesa_, incremento*getRarita()/(2*divisore));
        normalizza();
    }

    @Override
    public void potenzia(Double mana, String parametro) {
        Double incremento=mana/getLivello();
        Integer divisore =5;
        incrementStat(parametro,incremento);
        incrementStat(attacco_, incremento*getRarita()/(2*divisore));
        incrementStat(difesa_, incremento*getRarita()/(2*divisore));
        normalizza();
    }


}

File: tempIdeas/untitled/osso.cpp
#include "osso.h"

Osso::Osso(int livello, int rarita, double spirito, double attacco, double difesa) : Oggetto(livello, rarita, spirito), attacco_("Attacco"), difesa_("Difesa") {
    insertStat(attacco_,attacco);
    insertStat(difesa_, difesa);
}

double Osso::getAttacco() const {
    return getValoreStat(attacco_);
}

double Osso::getDifesa() const {
    return getValoreStat(difesa_);
}

Osso *Osso::clone() const {
    return new Osso(*this);
}

double Osso::ricicla() const {
    return calcolaMana() / 2 + getRarita() * ((getValoreStat(attacco_) + getValoreStat(difesa_))) / 2 ;

}

void Osso::potenzia(double mana, std::string parametro) {

    mana = sanitizeMana(mana);

    double incremento = mana / getLivello();
    int divisore;

    list<string> statsList = getListaStats();

    if((std::find(statsList.begin(), statsList.end(), parametro) == statsList.end())) {
        divisore = 10;
        incremento = incremento / statsList.size();
        for(auto i = statsList.begin(); i != statsList.end(); i++)
            incrementStat(*i, incremento);
    }
    else {
        divisore = 5;
        incrementStat(parametro, incremento);
    }

    incrementStat(attacco_, incremento * getRarita() / (2*divisore));
    incrementStat(difesa_, incremento * getRarita() / (2*divisore));

    normalizza();
}
